### PR TITLE
TimeTable: Fix /departures and /arrivals

### PR DESCRIPTION
**Pull request by antoine-de**
there was a problem on departure/arrival when multiple JourneyPattern were passing by the stoppoint (as show in test `dep_arr_filtering` in get_stop_time_tests.cpp)

fix http://jira.canaltp.fr/browse/NAVITIAII-1853

### CODE REVIEW

File: source/time_tables/get_stop_times.cpp [hunk 1]
```diff
@@ -46,47 +46,44 @@ std::vector<datetime_stop_time> get_stop_times(const routing::StopEvent stop_eve
...
-    auto test_add = true;
     routing::NextStopTime next_st = routing::NextStopTime(data);
 
-    // Next departure for the next stop: we use it not to have twice the same stop_time
-    // We init it with the wanted date time
-    std::map<routing::JppIdx, DateTime> next_requested_datetime;
-    for(auto jpp_idx : journey_pattern_points){
-        next_requested_datetime[jpp_idx] = dt;
-    }
+    // Next departure for the next stop: we store it to have to next departure for each jpp
```
Comment (pbougue) [inline source/time_tables/get_stop_times.cpp]: to have `the` next departure

File: source/time_tables/get_stop_times.cpp [hunk 2]
```diff
@@ -46,47 +46,44 @@ std::vector<datetime_stop_time> get_stop_times(const routing::StopEvent stop_eve
...
+    // We init it with the next_stop_time for each jpp
+    JppStQueue next_requested_dt({clockwise});
```
Comment (xlqian) [inline source/time_tables/get_stop_times.cpp]: you can remove the parenthesis?

File: source/time_tables/previous_passages.cpp [hunk 3]
```diff
@@ -71,7 +70,7 @@ previous_passages(const std::string &request,
 
     for(auto dt_stop_time : passages_dt_st) {
         pbnavitia::Passage * passage;
-        if(stop_event == StopEvent::pick_up) {
+        if(vis.api_pb == pbnavitia::PREVIOUS_DEPARTURES) {
             passage = handler.pb_response.add_next_departures();
```
Comment (pbougue) [inline source/time_tables/previous_passages.cpp:74]: Maybe rename that method to add_pick_up() and add_drop_off() for the other one.

File: source/time_tables/previous_passages.cpp [hunk 4]
```diff
@@ -71,7 +70,7 @@ previous_passages(const std::string &request,
 
     for(auto dt_stop_time : passages_dt_st) {
         pbnavitia::Passage * passage;
-        if(stop_event == StopEvent::pick_up) {
```
Comment (pbougue) [inline source/time_tables/previous_passages.cpp:74]: This change was made in prevision of a re-factor and merge with next_passages (divide previous/next on one side and pick_up/drop_off on another side).
The second reason for this is semantic : `pick_up` is far less ambiguous than `departure` (as one can think of trip departure in that case).
So some re-factor is still missing, but this was the direction I took (and I was maybe a little lazy to not do that, PR was already a little fat).

File: source/time_tables/previous_passages.cpp [hunk 5]
```diff
@@ -71,7 +70,7 @@ previous_passages(const std::string &request,
 
     for(auto dt_stop_time : passages_dt_st) {
         pbnavitia::Passage * passage;
```
Comment (pbougue) [inline source/time_tables/previous_passages.cpp]: spacing :)

File: source/time_tables/next_passages.cpp [hunk 6]
```diff
@@ -69,9 +68,9 @@ next_passages(const std::string &request,
     size_t total_result = passages_dt_st.size();
     passages_dt_st = paginate(passages_dt_st, count, start_page);
 
-    for(auto dt_stop_time : passages_dt_st) {
-        pbnavitia::Passage * passage;
-        if(stop_event == StopEvent::pick_up) {
```
Comment (pbougue) [inline source/time_tables/next_passages.cpp]: same as previous_passages

File: source/time_tables/tests/get_stop_times_test.cpp [hunk 7]
```diff
@@ -333,6 +333,26 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
...
+        const auto& vec_jpp_stop = b.data->dataRaptor->jpps_from_sp[sp_idx];
+        for (const auto& jpp : vec_jpp_stop) {
+            jpp_stop.push_back(jpp.idx);
+        }
+        return jpp_stop;
+    }
+
+    ed::builder b;
+    const navitia::DateTime yesterday = navitia::DateTimeUtils::min;
+    const navitia::DateTime yesterday_8h45 = navitia::DateTimeUtils::set(0, 8*3600 + 45*60);
```
Comment (pbougue) [inline source/time_tables/tests/get_stop_times_test.cpp]: use `_t` ?

File: source/time_tables/tests/get_stop_times_test.cpp [hunk 8]
```diff
@@ -333,6 +333,26 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
 
 }
 
+struct departure_helper {
```
Comment (pbougue) [inline source/time_tables/tests/get_stop_times_test.cpp:336]: :+1:

File: source/time_tables/tests/get_stop_times_test.cpp [hunk 9]
```diff
@@ -410,3 +414,142 @@ BOOST_AUTO_TEST_CASE(test_discrete_next_arrivals_prev_departures){
...
+ * s1-----------s2------------s3
+ *              |              |
+ *              |              |
+ *              ----------------
+ *
+ *
+ */
+BOOST_FIXTURE_TEST_CASE(test_discrete_lolipop, departure_helper) {
+    b.vj("A")("s1", "8:00"_t, "8:01"_t)
+            ("s2", "9:00"_t, "9:01"_t)
```
Comment (pbougue) [inline source/time_tables/tests/get_stop_times_test.cpp]: minor: indent

**Comment by pbougue** [comment]
Ok for me

**Pull request merged**